### DISCUSSION TITLE
1D ODE - first derivative term.

### DISCUSSION THREAD

**Discussion by ethrelfall** [firedrakeproject/firedrake#2688] [Firedrake support] [answered]
Dear community,

I want to specify the weak form of the 1D ODE $u' + (x u')' + 1 = 0$.

Once multiplied by a test function v and integrated by parts, the first term gives me $- \int u v' \text{d}x$.

How can I specify this term in Firedrake e.g. I cannot do `u*grad(v)` or `inner(u, grad(v))` due to shape mismatch (it thinks one of the objects is a vector and one a scalar)?  Note the second term in the original eq. is fine as `x*inner(grad(u), grad(v))` is allowed.

I guess one way around the problem is to put the ODE into self-adjoint-like form $(p u')'+q u = \text{stuff}$ ... but what if I don't want to do that?

(Checked for matching discussions.)

Thanks!

**Comment by wence-** [comment, marked as answer]
UFL doesn't exploit (kind of deliberately, I think) the isomorphism between scalars, and 1D vectors. So the way round this is that you need to index the `grad(v)` term (or equally just take the `x` derivative because you know that's the only one).

So try with either `u*grad(v)[0]` or `u*v.dx(0)`.

**Reply by ethrelfall** [reply]
Many thanks, that works fine!
